azure/services/bastionhosts/bastionhosts.go

```diff
@@ -38,6 +42,13 @@ func New(scope BastionScope) *aso.Service[*asonetworkv1.BastionHost, BastionScop
 	if spec != nil {
 		svc.Specs = []azure.ASOResourceSpecGetter[*asonetworkv1.BastionHost]{spec}
 	}
+	svc.ListFunc = list
 	svc.ConditionType = infrav1.BastionHostReadyCondition
 	return svc
 }
+
+func list(ctx context.Context, client client.Client, opts ...client.ListOption) ([]*asonetworkv1.BastionHost, error) {
+	list := &asonetworkv1.BastionHostList{}
+	err := client.List(ctx, list, opts...)
+	return slice.ToPtrs(list.Items), err
+}
```

Comment:
I would love for generics to smooth this out without having to redo this in each service, but there are a couple things that stand in the way there:
- There isn't a good way I know of to dynamically get this `BastionHostList` type given a type like `BastionHost`
- The `BastionHost`s can't be extracted generically from a `BastionHostList` since the `.Items` field of the list isn't exposed by any interface common to all list types that I can find. If anything, I think that would be this: https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListMeta